### PR TITLE
[SPARK-9114] [SQL] [PySpark] convert returned object from UDF into internal type

### PR DESCRIPTION
**Pull request by davies**
This PR also remove the duplicated code between registerFunction and UserDefinedFunction.

cc @JoshRosen

### CODE REVIEW

**Comment by davies** [comment]
@JoshRosen Is there a way to fix the MiMa checker to not complaint the missing of a local function inside another function?

**Comment by JoshRosen** [comment]
@davies, I think that MiMa complained here because you removed the the public `builder` method from the public `UDFRegistration` class.  Was this method not intended to be public?

**Comment by JoshRosen** [comment]
**Edit**: ignore earlier comment about MiMa; I think it's just a MiMa bug, so the workaround is fine.

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -658,23 +658,26 @@ class UserDefinedFunction(object):
...
-        self._judf = self._create_judf()
+        self._judf = self._create_judf(name)
 
-    def _create_judf(self):
-        f = self.func  # put it in closure `func`
-        func = lambda _, it: map(lambda x: f(*x), it)
+    def _create_judf(self, name):
+        f, returnType = self.func, self.returnType  # put them in closure `func`
+        from pyspark.serializers import CloudPickleSerializer
+        CloudPickleSerializer().dumps(returnType)
```
Comment (JoshRosen) [inline python/pyspark/sql/functions.py]: What is this line doing?  It looks like we're calling it only for side-effects.  Is it necessary?

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -658,23 +658,26 @@ class UserDefinedFunction(object):
...
+        func = lambda _, it: map(lambda x: returnType.toInternal(f(*x)), it)
```
Comment (JoshRosen) [inline python/pyspark/sql/functions.py]: Note for other reviewers: this is the key line of this patch, which performs the return type conversion; the rest of the patch is cleanup / refactoring.

**Comment by JoshRosen** [comment]
LGTM save for one question about a `CloudPickleSerializer` call; it looks like this fixes the return type conversion issue and also performs some nice refactoring / cleanup of duplicated code.

File: python/pyspark/sql/functions.py (hunk 1 above)
Comment (davies) [inline python/pyspark/sql/functions.py]: It's debugging code, will remove it.

**Pull request closed without merging**